### ISSUE TITLE
StorageOs Volume Provisioning in K8s Cluster - Issue

### ISSUE THREAD

**Issue by arjunm183** [storageos/helm-chart#6]
Hello Team,

Is this a BUG Report ? : YES

**Issue in Brief:** StorageOs volume create doesnt create volumes in the Hostpath, however it only creates inside the container.

We are facing an issue with StorageOs deployed in our kubernetes Environment ( Kubeadm 1.9.2 )

The Helm chart storageos deployment using daemonset has the option 
***************************************
- mountPath: /var/lib/storageos
  mountPropagation: Bidirectional
***************************************
**Pre-Requisites Done:**

Kubeadm manifests were modified  and restarted and restarted . It runs perfectly fine with --feature-gates=MountPropogation:True

**=====================kube-apiserver===========================**
root     10005  9945 10 Apr01 ?        03:48:20 kube-apiserver --insecure-port=0 --requestheader-extra-headers-prefix=X-Remote-Extra- --requestheader-allowed-names=front-proxy-client --service-account-key-file=/etc/kubernetes/pki/sa.pub --client-ca-file=/etc/kubernetes/pki/ca.crt --tls-cert-file=/etc/kubernetes/pki/apiserver.crt --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt --enable-bootstrap-token-auth=true --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname --requestheader-username-headers=X-Remote-User --tls-private-key-file=/etc/kubernetes/pki/apiserver.key --kubelet-client-certificate=/etc/kubernetes/pki/apiserver-kubelet-client.crt --proxy-client-key-file=/etc/kubernetes/pki/front-proxy-client.key --admission-control=Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,ResourceQuota --requestheader-group-headers=X-Remote-Group --proxy-client-cert-file=/etc/kubernetes/pki/front-proxy-client.crt --allow-privileged=true --advertise-address=52.244.203.197 --service-cluster-ip-range=10.96.0.0/12 --kubelet-client-key=/etc/kubernetes/pki/apiserver-kubelet-client.key --feature-gates=PersistentLocalVolumes=true,VolumeScheduling=true,MountPropagation=true --secure-port=6443 --authorization-mode=Node,RBAC --etcd-servers=http://127.0.0.1:2379
**=====================Scheduler===========================**
root     10184 10153  1 Apr01 ?        00:39:15 kube-scheduler --address=127.0.0.1 --leader-elect=true --feature-gates=PersistentLocalVolumes=true,VolumeScheduling=true,MountPropagation=true --kubeconfig=/etc/kubernetes/scheduler.conf

**=====================Controller Manager===========================**
root     11351 11333  5 Apr01 ?        01:49:02 kube-controller-manager --use-service-account-credentials=true --kubeconfig=/etc/kubernetes/controller-manager.conf --cluster-signing-key-file=/etc/kubernetes/pki/ca.key --address=127.0.0.1 --leader-elect=true --service-account-private-key-file=/etc/kubernetes/pki/sa.key --cluster-signing-cert-file=/etc/kubernetes/pki/ca.crt --controllers=*,bootstrapsigner,tokencleaner --root-ca-file=/etc/kubernetes/pki/ca.crt --allocate-node-cidrs=true --cluster-cidr=10.244.0.0/16 --node-cidr-mask-size=24 --feature-gates=PersistentLocalVolumes=true,VolumeScheduling=true,MountPropagation=true
root     31859 23027  0 12:58 pts/2    00:00:00 grep --color=auto controller
=============================================================

**Observation :** When i create a volume it says volume created. However i couldn't see the volume created in the HostPath /var/lib/storage, however i see the volume entry got created inside the container if i do a exec on the docker container running storageos/node 

Output Below
**==========storageos volume list===============**
storageos volume list
NAMESPACE/NAME                                     SIZE                MOUNTED BY          NODE SELECTOR       STATUS              REPLICAS            LOCATION
default/pvc-0ca6c5a6-3635-11e8-9ed3-005056bb362d   5GB                                                         active              0/0                 sgscall061.inedc.corpintra.net (healthy)

**===========docker exec storageos container==============**

[root@sgscall061 ~]# docker exec -it 1a0be9c9c1ce sh
# cd /var/lib/storageos/
# ls
data  kv  state  volumes
# cd volumes
# ls
**338dcfd3-f9a7-4fba-c982-f5fe084f8231**  a5d62846-68c7-3096-8e16-111cee079706

**===========HostPath of the machine ==============**
[root@sgscall061 volumes]# pwd
/var/lib/storageos/volumes
[root@sgscall061 volumes]# ls -lah
total 0
drwx------. 2 root root  6 Apr  1 01:31 .
drwxr-xr-x. 6 root root 56 Apr  1 01:31 ..

No Volumes Present.

**Error:**
https://docs.storageos.com/docs/install/schedulers/kubernetes

As mentioned in the above link i get the error when i create the pod.
Can you help me in identifying and fixing the issue regarding this.
I can see you have suggested to add "--volume=/var/lib/storageos:/var/lib/storageos:rshared "

But this is for the docker environment. Can you let me know how do i add those options with the kubernetes environment.( I beleive thats the reason HostPath and MountPropogation is mentioned in the daemonset )

Looks like a bug for me.

**Comment by arjunm183** [comment]
Hello Team,

Issue was fixed by adding kubelet Args with feature-gates in it.

Thanks and Regards,
Arjun.M

**Comment by oicheryl** [comment]
No worries @arjunm183, let us know if you have any other issues.

**Comment by arjunm183** [comment]
@oicheryl Thanks. One question why the daemonset counts are (3) . When i deployed daemonset it was running only on 3 nodes. Is it a limitation? How do we increase this  to all the k8s nodes.

**Comment by oicheryl** [comment]
HI @arjunm183, how many nodes are you deploying on? The daemonset should automatically scale it to all the Kubernetes worker nodes.

**Comment by arjunm183** [comment]
@oicheryl I have 5 nodes including the master node.
However daemonset did not ran on master and one more worker node.

**Comment by oicheryl** [comment]
It's not (intentionally) limited to 3 nodes, it should run on the fourth worker node, but not the master node.

Where did you see it only running on 3 nodes? Are there any events on the fourth node to show if the pod was started and crashed?

**Comment by arjunm183** [comment]
i had pulled storageos cli binary and checked it shows only three worker nodes.
( storageos node ls ) --> shows only three worker nodes.
I couldn't see the fourth node having the pod started message or crashed container logs( docker ps -a )

**Comment by arjunm183** [comment]
@oicheryl I am really sorry for the trouble. I fixed it without reploying it. It was because i had made my node taint with value node-role.kubernetes.io/master:NoSchedule on one of the workernode. 
I untainted the node now.  This solved the issue. THanks a ton. 

One more question : Incase if we have to arrange a call with StorageOS , is it possible. 
Are there any channels ( slack ) where we can ask for help and support.
